c10d: add Collectives abstraction

d4l3k opened the pull request:
This adds a new `Collectives` API for doing distributed collectives operations. This is intended to replace the [current Elastic store abstraction](https://github.com/pytorch/pytorch/blob/main/torch/distributed/elastic/utils/store.py) with more performant and debugable primitives.

Design doc: https://docs.google.com/document/d/147KcKJXEHvk1Q6tISLbJVvLejHg_1kIhBQeu-8RQxhY/edit

The standard implementation is using `StoreCollectives` but other more performant backends will be added in a follow up PR.

Test plan:

```
python test/distributed/test_collectives.py -v
```

This tests both functionality using multiple threads as well as timeout behavior.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

d4l3k committed:
c10d: added Collectives abstraction

shuqiangzhang commented on torch/csrc/distributed/c10d/Collectives.hpp:
```diff
@@ -0,0 +1,59 @@
+class TORCH_API Collectives : public torch::CustomClassHolder {
+ public:
+  virtual void barrier(
+      const std::string& prefix,
```
Is 'prefix' specific to store based implementation of the APIs? If so, I imagine the APIs could be implemented with non-store semantics

d4l3k replied:
It refers to the store based APIs but other implementations have a similar concept of an operation ID. I.e. the label field in `https://fburl.com/code/dq8shbwi`.

It may be best to change this name from prefix to "operation_uid"

d4l3k replied:
I think this is still a better fit for label -- all operations on store must be sequential regardless of prefix.

If you want a fully separate namespace you can use a different PrefixStore and create a new StoreCollectives instance -- similar to how you would use a different sub_world_id and create a new SCCollectives instance

d4l3k replied:
Renamed to generic `key`

kwen2501 commented:
Does the term `Collectives` here refer to "collectives for control plane (torchrun/elastic)"? 
If so, can file name and class name like `Collectives.hpp` be renamed to more clearly differentiate from the data plane collectives (i.e. c10d)? Thanks!

kwen2501 commented:
In addition, it seems torch/csrc/distributed/c10d has many files now. Does it make sense for store to have its own directory? I am just not sure whether the new directory should under c10d or at the same level of c10d. If store is used by both c10d and torchrun/elastic, it seems the latter makes sense. However, a lot of store files also have c10d namespace, so that makes me uncertain.

Skylion007 commented on torch/csrc/distributed/c10d/StoreCollectives.cpp:
```diff
@@ -0,0 +1,191 @@
+    c10::intrusive_ptr<::c10d::Store> store,
+    int rank,
+    int world_size)
+    : store_(store), rank_(rank), world_size_(world_size) {}
```
```suggestion
    : store_(std::move(store)), rank_(rank), world_size_(world_size) {}
```

Skylion007 commented on torch/csrc/distributed/c10d/StoreCollectives.cpp:
```diff
@@ -0,0 +1,191 @@
+    std::chrono::milliseconds timeout) {
+  StoreTimeoutGuard g{*store_, timeout};
+
+  std::vector<std::string> keys;
```
```suggestion
  std::vector<std::string> keys;
  keys.reserve(world_size_);
```

Skylion007 commented on torch/csrc/distributed/c10d/StoreCollectives.cpp:
```diff
@@ -0,0 +1,191 @@
+    std::chrono::milliseconds timeout) {
+  StoreTimeoutGuard g{*store_, timeout};
+
+  std::vector<std::string> keys;
```
```suggestion
  std::vector<std::string> keys;
  keys.reserve(world_size_);
```

Skylion007 commented on torch/csrc/distributed/c10d/StoreCollectives.cpp:
```diff
@@ -0,0 +1,191 @@
+  auto local_key = prefix + "/" + std::to_string(rank_);
+  store_->set(local_key, data);
+
+  std::vector<std::string> keys;
```
```suggestion
  std::vector<std::string> keys;
  keys.reserve(world_size_);
```

Skylion007 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1487,6 +1489,195 @@ that adds a prefix to each key inserted to the store.
+                std::vector<std::vector<uint8_t>> out =
+                    collectives.all_gather(prefix, data_, timeout);
+
+                std::vector<std::string> out_;
```
```suggestion
                std::vector<std::string> out_;
                out_.reserve(out.size());
```

Skylion007 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1487,6 +1489,195 @@ that adds a prefix to each key inserted to the store.
+                 const std::string& prefix,
+                 const std::vector<std::string>& data,
+                 std::chrono::milliseconds timeout = 5min) {
+                std::vector<std::vector<uint8_t>> data_;
```
```suggestion
                std::vector<std::vector<uint8_t>> data_;
                data_.reserve(data.size());
```

Skylion007 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1487,6 +1489,195 @@ that adds a prefix to each key inserted to the store.
+                std::vector<std::vector<uint8_t>> out =
+                    collectives.gather_recv(prefix, data_, timeout);
+
+                std::vector<std::string> out_;
```
```suggestion
                std::vector<std::string> out_;
                out_.reserve(out.size());
```

Skylion007 reviewed (commented):
Performance nits / improvements.

Skylion007 commented on torch/csrc/distributed/c10d/StoreCollectives.cpp:
```diff
@@ -0,0 +1,191 @@
+  auto last_members_key = prefix + "/last_members";
+
+  auto idx = store_->add(num_members_key, 1);
+  store_->set(prefix + "/" + std::to_string(rank_), "joined");
```
since you are doing a lot of string concatenation in these files, may want to consider looking to see if fmtlib help makes the code simpler + faster. It's already a dependency in PyTorch.

d4l3k commented:
updates:

* moved to c10d/control_collectives dir
* renamed Collectives to ControlCollectives
* renamed `prefix` to `key` to be backend agnostic
* applies fmt/vector reserve optimizations

kurman commented on torch/csrc/distributed/c10d/control_collectives/StoreCollectives.hpp:
```diff
@@ -0,0 +1,62 @@
+
+namespace c10d {
+
+class TORCH_API StoreCollectives : public ControlCollectives {
```
StoreControlCollectives?

d4l3k replied:
I thought about that but figured Store would be clear enough -- it wouldn't make sense to use the Store API for dataplane operations

kurman commented on torch/csrc/distributed/c10d/control_collectives/ControlCollectives.hpp:
```diff
@@ -0,0 +1,59 @@
+class TORCH_API ControlCollectives : public torch::CustomClassHolder {
+ public:
+  virtual void barrier(
+      const std::string& key,
```
`key` is better, I wonder if it is actually a 'token'?

d4l3k replied:
I think they're equivalent -- key maps better to the underlying concept so will just leave it as is

kurman reviewed (commented):
nice

shuqiangzhang commented on torch/csrc/distributed/c10d/Store.hpp:
```diff
@@ -97,4 +97,33 @@ class TORCH_API Store : public torch::CustomClassHolder {
+
+ private:
+  Store& store_;
+  std::chrono::milliseconds old_timeout_;
```
nit: do we have a recommended format style in Pytorch for member variables?  saw some mixed used of lowerCamelCase_ and this style in the code base.

d4l3k replied:
TIL https://fburl.com/wiki/qge4qo0c

looks like `lowerCamelCase_` is the one to pick -- my bad, will update

d4l3k replied:
done

shuqiangzhang commented on torch/csrc/distributed/c10d/control_collectives/StoreCollectives.cpp:
```diff
@@ -0,0 +1,205 @@
+}
+
+void StoreCollectives::broadcast_send(
+    const std::string& key,
```
For these simple APIs such as broadcast, is 'key' really necessary for users? can we make it optional for users? e.g., they just need to provide the data to send and recv

d4l3k replied:
@shuqiangzhang we need a unique identifier for every round due to some implementation details -- we could auto increment it internally but could cause problems. SC and store backends both require a unique key

shuqiangzhang commented on test/distributed/test_control_collectives.py:
```diff
@@ -0,0 +1,159 @@
+
+        def f(rank: int) -> None:
+            collectives = dist.StoreCollectives(store, rank, world_size)
+            collectives.barrier("foo", timedelta(seconds=10), True)
```
Yeah, I mean this 'foo' does not really provide much value for users? could be optional?

d4l3k replied:
For store, this needs to be unique, if you call it multiple times with "foo" it won't actually do any barrier operation

shuqiangzhang replied:
OK, I was thinking it would be much nicer and cleaner if our general API could remove the burden from users of maintaining an unique key for each collective due to implementation limitations. Can we maintain a seqID or something within the implementation?

d4l3k replied:
we could potentially -- it gets a bit risky when you consider the case that not all workers may be joining at the same time. In elastic it's elastic so a simple seqID may not be sufficient and you need to key it off of the rendezvous round instead

shuqiangzhang replied:
OK, this makes sense in elastic

shuqiangzhang reviewed (approved):
LGTM, great start of the new control plane!  Please do remember to unify the variable naming scheme

d4l3k commented:
@pytorchbot merge

d4l3k commented:
@pytorchbot merge

d4l3k commented:
@pytorchbot merge

DanilBaibak commented:
@pytorchbot revert -m "Break internal build" -c ghfirst

DanilBaibak commented:
Hi @d4l3k! Sorry, I need to revert your PR because it breaks the internal build. Here you can find more information about the issue - [D57518638](https://www.internalfb.com/diff/D57518638).
